kernel/Utilities/ElementLocalIndexing.h

```diff
@@ -0,0 +1,152 @@
+    /// than the maximum.
+    ///
+    /// \param includedUnknowns The new set of included unknowns.
+    void reinit(const std::vector<std::size_t>& includedUnknowns);
+    /// Reinit the values for a different element
+    /// \param element The new element (may be nullptr)
+    void reinit(const Base::Element* element);
+
+    /// Combined reinit for both the element and included unknowns.
+    ///
+    /// \param element The new element (may be nullptr)
+    /// \param includedUnknowns The new set of included unknowns.
+    void reinit(const Base::Element* element,
+                const std::vector<std::size_t>& includedUnknowns) {
+        reinit(includedUnknowns);
+        reinit(element);
+    }
+
+    /// Get the Offset in an ElementMatrix/Vector for the unknown. Will be an
+    /// invalid value for non included unknowns or when no element is available.
+    std::size_t getDoFOffset(std::size_t unknown) const {
+        logger.assert_debug(unknown < numberOfUnknowns_,
+                            "Unknown % larger than maximum %", unknown,
+                            numberOfUnknowns_);
+        return offsets_[unknown];
+    }
+    /// Get the number DoFs (rows/columns) for the unknown in an element
+    /// matrix/vector. Will be zero for non included unknowns or when no element
+    /// is available.
+    std::size_t getNumberOfDoFs(std::size_t unknown) const {
+        logger.assert_debug(unknown < numberOfUnknowns_,
+                            "Unknown % larger than maximum %", unknown,
+                            numberOfUnknowns_);
+        return sizes_[unknown];
+    }
+    /// Total number of DoFs of the included unknowns and thus rows/columns of
+    /// an element matrix or vector. Without element this will be zero.
+    std::size_t getNumberOfDoFs() const { return totalNumberOfDofs_; }
+
+    /// Get the current element.
+    const Base::Element* getElement() const { return element_; }
+    /// The unknowns that are included.
+    const std::vector<std::size_t>& getIncludedUnknowns() const {
+        return includedUnknowns_;
+    }
+
+    /// Check the invariants of the datastructure to see if the internal
+    /// constraints are satisfied. To be used for testing and debugging.
+    void validate() const;
+
+    /// Constructor alternative for an index where all unknowns are included.
+    /// \param numberOfUnknowns The total number of unknowns
+    /// \return A new instance with all unknowns included
+    static ElementLocalIndexing createFullIndexing(
+        std::size_t numberOfUnknowns) {
+        std::vector<std::size_t> unknowns(numberOfUnknowns);
+        std::iota(unknowns.begin(), unknowns.end(), 0);
+        ElementLocalIndexing result(numberOfUnknowns);
+        result.reinit(unknowns);
+        return result;
+    }
+
+   private:
+    /// The current element
+    const Base::Element* element_;
+
+    /// The maximum number of unknowns.
+    std::size_t numberOfUnknowns_;
```

Comment:
is that not simply `sizes_.size()`

Reply:
Yes, and also to `offsets_.size()`, but both are far less descriptive in my opinion.

Reply:
It's now gone entirely and the size of `offsets_` and `sizes_` is now dynamical (i.e. sufficient to for the largest included unknown).